### PR TITLE
[v7r1] Fix jobStatusBulk

### PR DESCRIPTION
**Pull request by phicharp**
BEGINRELEASENOTES
*WorkloadManagement
FIX: There was a bug handling the bulk status updates that was not updating correctly the job status although the job logging information was correct). This PR fixes this bug. It was tested in certification as a hotfix
CHANGE: internally always call the `setJobStatusBulk()` method in order to be consistent
CHANGE: in `JobReport`, only send statuses that are not empty, and remove handling of `ApplicationCounter` that is never sent nor used anywhere

ENDRELEASENOTES

### CODE REVIEW

**Comment by phicharp** [comment]
In order to make sure we don't have time rounding problems, I am tempted to remove the time rounding in `JobLoggingDB.py`, method `addLoggingRecord()` as this rounding doesn't seem justified and may lead to confusion when sorting records. The rounding is at the 1/100 second level but i see no reason as we anyway store a floating point...

File: WorkloadManagementSystem/Service/JobStateUpdateHandler.py [hunk 1]
```diff
@@ -183,27 +183,38 @@ def export_setJobStatusBulk(self, jobID, statusDict):
     lastTime = Time.toString(Time.fromEpoch(lastTime))
 
     dates = sorted(statusDict)
+    log = self.log.getSubLogger('JobStatusBulk/Job-%s' % jobID)
+    log.verbose("*** New call *** Last update time %s - Sorted new times %s" % (lastTime, dates))
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/JobStateUpdateHandler.py:187]: Please make this a `debug`, or change it to `log.verbose("Setting bulk status", "Last update time %s - Sorted new times %s" % (lastTime, dates))`

File: WorkloadManagementSystem/Service/JobStateUpdateHandler.py [hunk 2]
```diff
@@ -183,27 +183,38 @@ def export_setJobStatusBulk(self, jobID, statusDict):
...
       if minor == "Application" and status == JobStatus.RUNNING and not startTime:
         startTime = date
 
     # We should only update the status if its time stamp is more recent than the last update
     if dates[-1] >= lastTime:
       # Get the last status values
-      for date in [date for date in dates if date >= lastTime]:
+      for date in [dt for dt in dates if dt >= lastTime]:
         sDict = statusDict[date]
+        log.verbose("\tTime %s - Statuses %s" % (date, str(sDict)))
```
Comment (fstagni) [inline WorkloadManagementSystem/Service/JobStateUpdateHandler.py:209]: Same here. I would prefer `debug`

Comment (phicharp) [reply WorkloadManagementSystem/Service/JobStateUpdateHandler.py:209]: OK, will do... I set it verbose because the service is running in verbose mode ;-) 
`log` is already a sub logger with `jobStatusBulk`...

Comment (phicharp) [reply WorkloadManagementSystem/Service/JobStateUpdateHandler.py:209]: I thought the rule `(<fixed string>, <variable string>)` was only mandatory for levels more severe than INFO?

Comment (fstagni) [reply WorkloadManagementSystem/Service/JobStateUpdateHandler.py]: Also for `VERBOSE`.

**Comment by fstagni** [comment]
OK for removing the rounding.

File: WorkloadManagementSystem/Service/JobStateUpdateHandler.py (hunk 2 above)
Comment (phicharp) [reply WorkloadManagementSystem/Service/JobStateUpdateHandler.py]: Done anyway

File: WorkloadManagementSystem/DB/JobLoggingDB.py [hunk 3]
```diff
@@ -50,38 +50,27 @@ def addLoggingRecord(self,
...
+        # The date is provided as a string in UTC
+        _date = Time.fromString(date)
+      elif isinstance(date, Time._dateTimeType):
+        _date = date
+      else:
+        self.log.error('Incorrect date for the logging record')
         _date = Time.dateTime()
-        epoc = time.mktime(_date.timetuple()) - MAGIC_EPOC_NUMBER
-        time_order = round(epoc, 3)
+    except BaseException:
```
Comment (fstagni) [inline WorkloadManagementSystem/DB/JobLoggingDB.py:65]:
```suggestion
    except Exception:
```

Comment (phicharp) [reply WorkloadManagementSystem/DB/JobLoggingDB.py]: Done but this makes `pylint` unhappy ;-)

**Comment by phicharp** [comment]
To reduce information sent to the server, I will remove the empty items in the dictionary (in `sendStoredStatusInfo()`).
Also I notice that `ApplicationCounter` is not set, hence doesn't need to be handled in the service, does it?

File: WorkloadManagementSystem/Client/JobReport.py [hunk 4]
```diff
@@ -132,20 +137,12 @@ def sendStoredJobParameters(self):
...
-    parameters = []
-    for pname, value in self.jobParameters.items():
-      pvalue, _timeStamp = value
-      parameters.append([pname, pvalue])
-
+    parameters = [[pname, value[0]] for pname, value in self.jobParameters.items()]
     if parameters:
       result = JobStateUpdateClient().setJobParameters(self.jobID, parameters)
-      if not result['OK']:
-        return result
```
Comment (fstagni) [inline WorkloadManagementSystem/Client/JobReport.py:143]: Why did you remove these 2 lines above?

Comment (phicharp) [reply WorkloadManagementSystem/Client/JobReport.py:143]: because there is a `return result` below

**Review by fstagni — APPROVED** [review]

**Comment by atsareg** [comment]
I would keep the ApplicationCounter. JobReporter is not necessaily the only interface to update the job status

**Comment by fstagni** [comment]
> JobReporter is not necessaily the only interface to update the job status

`JobReport` is right now the only client of `JobStateUpdateClient().setJobStatusBulk`

**Comment by phicharp** [comment]
This method is essentially a shortcut for setting the 3 statuses (as job attributes) at once and adding it to the job logging info, all in one call. Currently there is no attribute `ApplicationCounter` set or used anywhere. We have no clue why this had been foreseen... If necessary, a `setJobAttribute()` service method can be added to the `JobManagerHandler` that could add any attribute to a job.

**Comment by phicharp** [comment]
I have another question: `setJobStatus()` is not as well protected against delayed calls (e.g. from a failover request). I was wondering whether it would not make sense to call `setJobStatusBulk()` from within `setJobStatus()` to be equally protected when that method is used. This would introduce a small overhead as there are a few more checks but as anyway most of the time the bulk method is used already, even for a single update.... ???

**Comment by fstagni** [comment]
Yes, possible. In practice, IIUC we should one day have **only** `setJobStatusBulk`.
And since you are there you can mark `setJobsStatus` (notice the extra 's') as deprecated as looks unused to me.

**Comment by phicharp** [comment]
OK, I'll look into doing that...

**Comment by phicharp** [comment]
Oups, the target branch was incorrect, sorry...

**Comment by phicharp** [comment]
Are these failures meaningful? I don't understand anything to the log, sorry...

**Comment by fstagni** [comment]
They are yes. You can see this line? https://github.com/DIRACGrid/DIRAC/pull/4839/checks?check_run_id=1453158997#step:8:991

I anyway copy-paste below:

```
======================================================================
FAIL: test_FullChain (__main__.WMSChain)
This test will
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 127, in test_FullChain
    self.assertTrue(res['OK'], res.get('Message'))
AssertionError: MySQL Error ( 1131 : Execution failed.: ( 1292: Incorrect datetime value: 'None' for column 'EndExecTime' at row 1 ))

======================================================================
FAIL: test_ParametricChain (__main__.WMSChain)
This test will submit a parametric job which should generate 3 actual jobs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 201, in test_ParametricChain
    self.assertTrue(res['OK'], res.get('Message'))
AssertionError: MySQL Error ( 1131 : Execution failed.: ( 1292: Incorrect datetime value: 'None' for column 'EndExecTime' at row 1 ))

======================================================================
FAIL: test_JobStateUpdateAndJobMonitoring (__main__.JobMonitoring)
Verifying all JobStateUpdate and JobMonitoring functions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 242, in test_JobStateUpdateAndJobMonitoring
    self.assertTrue(res['OK'], res.get('Message'))
AssertionError: MySQL Error ( 1131 : Execution failed.: ( 1292: Incorrect datetime value: 'None' for column 'EndExecTime' at row 1 ))

======================================================================
FAIL: test_JobStateUpdateAndJobMonitoringMultuple (__main__.JobMonitoringMore)
# Now, let's submit some jobs. Different sites, types, inputs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 381, in test_JobStateUpdateAndJobMonitoringMultuple
    self.assertTrue(sorted(res['Value']) in [['Received'], sorted(['Received', 'Waiting'])])
AssertionError: False is not true

======================================================================
FAIL: test_matcher (__main__.Matcher)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 505, in test_matcher
    self.assertTrue(res['OK'], res.get('Message'))
AssertionError: MySQL Error ( 1131 : Execution failed.: ( 1292: Incorrect datetime value: 'None' for column 'EndExecTime' at row 1 ))

----------------------------------------------------------------------
Ran 6 tests in 7.917s

FAILED (failures=5)
```

**Comment by phicharp** [comment]
Thanks Federico, as I can't see any of those :(

**Comment by phicharp** [comment]
In `getJobAttributes()` all values are converted to strings, i.e. missing values are returned as `'None'` and not `None`. Hence some ugly code...
Would it be possible to not convert to strings?
Now waiting for the web site to work :(

**Comment by fstagni** [comment]
There is still something wrong:
```

======================================================================
FAIL: test_JobStateUpdateAndJobMonitoring (__main__.JobMonitoring)
Verifying all JobStateUpdate and JobMonitoring functions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 258, in test_JobStateUpdateAndJobMonitoring
    self.assertEqual(res['Value'], 'Running', msg="Got %s" % str(res['Value']))
AssertionError: Got Matched

======================================================================
FAIL: test_JobStateUpdateAndJobMonitoringMultuple (__main__.JobMonitoringMore)
# Now, let's submit some jobs. Different sites, types, inputs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dirac/ServerInstallDIR/DIRAC/tests/Integration/WorkloadManagementSystem/Test_Client_WMS.py", line 354, in test_JobStateUpdateAndJobMonitoringMultuple
    self.assertTrue(set(res['Value']) <= {'ANY', 'DIRAC.Jenkins.ch'}, msg="Got %s" % res['Value'])
AssertionError: Got ['ANY', 'DIRAC.Jenkins.ch', 'Site']

----------------------------------------------------------------------
Ran 6 tests in 15.007s

FAILED (failures=2)
```

**Comment by phicharp** [comment]
Ach, bloody test for testing a bug (job stuck in `Matched` status)... This now should not happen but was part of `setJobStatus` but not in bulk!

**Comment by phicharp** [comment]
Still trying to understand the second failure... I don't see how `'Site'` can come as a result of `getSites()` 
AFAIR there was no change in there...

**Comment by fstagni** [comment]
In Test_Client_WMS.py there's a call to `res = jobStateUpdate.setJobSite(jobID, 'Site')` that clearly would add that "'Site'". I don't know how it was not failing before. Not very important. This should hopefully be the last fix.

**Comment by phicharp** [comment]
Yes, this is what I had concluded, thus I have added `'Site'` to the list for success!

**Review by fstagni — APPROVED** [review]

**Review by atsareg — APPROVED** [review]
